barcode_utils.py

```diff
@@ -0,0 +1,166 @@
+from barcode import *
+
+import base64
+import struct
+import skip32
+import code128
+import os
+
+
+# TODO: currently broken, need to rewrite in the context of uber
+def generate_barcode_csv(filename):
+    badge_types = self.badge_types['badge_types']
+
+    # TODO: modify so we read the config directly from uber.
+    lines = []
+    for badge_type, ranges in badge_types.items():
+        range_start = int(ranges['range_start'])
+        range_end = int(ranges['range_end'])
+        lines = lines + self.generate_barcode_nums(range_start, range_end)
+
+    f = open(filename,'w')
+
+    for line in lines:
+        f.write(line + os.linesep)
+
+    f.close()
+
+
+# TODO: currently broken, need to rewrite in the context of uber
+def generate_barcode_nums(range_start, range_end):
+    generated_lines = []
+    seen_barcodes = []
+    for badge_num in range(range_start, range_end+1):
+        barcode_num = generate_barcode_from_badge_num(
+            badge_num=int(badge_num),
+            event_id=self.event_id,
+            salt=self.salt,
+            key=self.secret_key
+        )
+
+        line = "{badge_num},{barcode_num}".format(
+            badge_num=badge_num,
+            barcode_num=barcode_num,
+        )
+        generated_lines.append(line)
+
+        # ensure that we haven't seen this value before
+        if barcode_num in seen_barcodes:
+            raise ValueError('COLLISION: generated a badge# that\'s already been seen')
+        else:
+            seen_barcodes.append(barcode_num)
+
+    return generated_lines
+
+
+def generate_barcode_from_badge_num(badge_num, event_id=None, salt=None, key=None):
+    event_id = config['secret']['barcode_event_id'] if not event_id else event_id
+    salt = config['secret']['barcode_salt'] if not salt else salt
+    key = bytes(config['secret']['barcode_key'],'ascii') if not key else key
+
+    # packed data going to be encrypted is:
+    # byte 1 - 8bit event ID, usually 1 char
+    # byte 2,3,4 - 24bit badge number
+
+    salted_val = badge_num + (0 if not salt else salt)
+
+    if salted_val > 0xFFFFFF:
+        raise ValueError("badge_number is too high " + str(badge_num))
+
+    data_to_encrypt = struct.pack('>BI', event_id, salted_val)
+
+    # remove the highest byte in that integer (2nd byte)
+    data_to_encrypt = bytearray([data_to_encrypt[0], data_to_encrypt[2], data_to_encrypt[3], data_to_encrypt[4]])
+
```

Comment:
Minor stylistic note: in cases like this I generally just say

```python
assert len(data_to_encrypt) == 4 and len(key) == 10
```

and if you want to actually get fancy with a custom error message (which is probably not necessary, especially since you can have configobj enforce the key length of 10), you can say

```python
assert len(data_to_encrypt) == 4 and len(key) == 10, 'data_length={} key_length={}'.format(len(data_to_encrypt), len(key))
```